dal/src/main/java/com/epam/esm/dal/impl/CertificateDaoSql.java

```diff
@@ -0,0 +1,154 @@
+package com.epam.esm.dal.impl;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.sql.Types;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DataAccessException;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.PreparedStatementCreator;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+
+import com.epam.esm.dal.CertificateDao;
+import com.epam.esm.dal.TagDao;
+import com.epam.esm.dal.mapper.CertificateMapper;
+import com.epam.esm.dal.mapper.CertificateWithTagsMapper;
+import com.epam.esm.dal.util.DuplicateResultsRemover;
+import com.epam.esm.dal.util.SqlQueryBuilder;
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.Tag;
+import com.epam.esm.transferobj.FilterParam;
+import com.epam.esm.transferobj.OrderParam;
+
+@Repository
+public class CertificateDaoSql implements CertificateDao {
+
+	private final JdbcTemplate jdbcTemplate;
+	private SqlQueryBuilder sqlBuilder;
+	private TagDao tagDao;
+
+	private static final String sqlFindCertificateById = "SELECT * FROM GiftService.GiftCertificate WHERE Id = (?)";
+	private static final String sqlAddCertificate = "INSERT INTO GiftService.GiftCertificate (Name, Description, "
+			+ "Price, CreateDate, LastUpdateDate, Duration) VALUES (?,?,?,?,?,?);";
+	private static final String sqlInsertIntoM2M = "INSERT INTO GiftService.`Tag-Certificate` VALUES (?,?);";
+	private static final String sqlUpdateCertificate = "Update GiftService.GiftCertificate set Name = (?), "
+			+ "Description = (?), Price = (?), LastUpdateDate = (?), Duration = (?) where Id = (?);";
+	private static final String sqlDeleteCertificateById = "DELETE FROM GiftService.GiftCertificate WHERE Id = (?);";
+	private static final String sqlDeleteFromM2M = "DELETE FROM GiftService.`Tag-Certificate` WHERE IdCertificate = (?);";
+
+	@Autowired
+	public CertificateDaoSql(JdbcTemplate jdbcTemplate) {
+		this.jdbcTemplate = jdbcTemplate;
+	}
+
+	@Autowired
+	public void setTagDao(TagDao tagDao) {
```

Comment:
Could you please explain what is the reason not to inject TagDao with constructor?